### PR TITLE
Add Tizen backend

### PR DESCRIPTION
**Pull request by rookiejava**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Feature

**What is the current behavior? (You can also link to an open issue here)**
There is no tizen as a backend platfom of ReactiveUI

**What is the new behavior (if this is a feature change)?**
Tizen has been added as a backend platform  of ReactiveUI

**What might this PR break?**
None

**Please check if the PR fulfills these requirements**
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Other information**:
This PR is a latest revision of #1387. And, this PR contains only the minumum requrirements to support Tizen. So, implementations that need to work correctly (like android, iOS), should be required through different PRs.

### CODE REVIEW

**Comment by ghuntley** [comment]
Howdy @rookiejava thank-you so much for your contribution. I've noticed on your github profile that you have done some work on the tizen platform before. How well do you know the platform? Do you work at Samsung? We are looking for a platform lead. I'm happy to work with you, pair with you and video call with you to discuss the broader picture. Last I was aware was that .net ++ tizen was immature and it wasn't possible to actually target the platform with nuget (even though the tfm is defined) - has this changed?

**Comment by rookiejava** [comment]
@ghuntley Yes. The latest vesrion of Tizen (Tizen 4.0 M2), that offially supports .NET, was released in  the last month. For more details, see https://developer.tizen.org/tizen/release-notes/tizen-4.0-public-m2.

**Comment by ghuntley** [comment]
@rookiejava and @ideallee we are looking for two people (you?!) to join the core team of ReactiveUI and help shape the future of Tizen. See https://reactiveui.net/team/ for specifics and if this is of interest to please join us over in https://reactiveui.net/slack at https://reactivex.slack.com/archives/CAK67UA7M/p1526040674000046

Tizen is something we really want to add in Splat, ReactiveUI and Akavache. Let's do it (soon - like next release?)

**Comment by ghuntley** [comment]
This branch has been updated and is almost ready for merge. I'm having some difficulties with EventGenerator and tizen.  Any help would be appreciated. Parking this for now.

Repro:
> ./build.cmd

`src/EventBuilder/`  is the code generator for `src/ReactiveUI.Events/Events_TIZEN.cs`

**Comment by rookiejava** [comment]
@ghuntley As for this issue, I and my colleague (@WonyoungChoi) took a look inside what happend. And we found there is a bug on `EventTenplateInformation` class. It seems that it doesn;t work correctly in case of genereic instance type. 

https://github.com/reactiveui/ReactiveUI/blob/10b77971d2dc71b36ef051e94655c49fcc9edb8f/src/EventBuilder/Cecil/EventTemplateInformation.cs#L39-L47

According to above logic, if real type name contains kvp name ('T', 'U', 'R'), it replaced into real type name.  As a result, the below 2 cases replaced unintended.

-  IObservable<Tizen.NUI.UIComponents.Slider.MarkReachedEventArgs> MarkReached
   - `MarkReachedEventArgs` turned into `MarkBooleaneachedEventArgs`  
     -  (R => Boolean)

- Observable<Tizen.NUI.ImfManager.EventReceivedEventArgs> 
   - `EventReceivedEventArg` turned into `EventTizen.NUI.ImfManager.EventeceivedEventArg` 
     -  (R => Tizen.NUI.ImfManager.Event)

This is not a bug on only A, but everywhere.  So, I'd like to recomend 

a) fix EventTemplateInformation correctly for generic instance type
b) ignore these 2 exceptional cases as for now (even if temporaily way)

**Comment by rookiejava** [comment]
I've tweaked it a little bit and build successfully.

**Comment by swasdppd** [comment]
amazeeeeeeeed folks 99% who can hit the last onwe _?

File: src/EventBuilder/Cecil/EventTemplateInformation.cs [hunk 1]
```diff
@@ -17,12 +17,34 @@ public static class EventTemplateInformation
...
             {"Windows.Foundation.EventHandler`1", "EventHandler"},
-            {"Windows.Foundation.EventHandler`2", "EventHandler"}
+            {"Windows.Foundation.EventHandler`2", "EventHandler"},
+
+            // tizen40
+            {"System.Boolean", "Boolean"},
+            {"System.Boolean`1", "Boolean"},
+ 
+
+            {"System.EventHandler", "EventHandler"},
```
Comment (olevett) [inline src/EventBuilder/Cecil/EventTemplateInformation.cs]: I'm not sure the tizen40 comment is that clear, or necessarily adds that much

File: src/EventBuilder/Cecil/EventTemplateInformation.cs [hunk 2]
```diff
@@ -17,12 +17,34 @@ public static class EventTemplateInformation
...
+
+            {"System.EventArgs", "EventArgs"},
+            {"System.EventArgs`1", "EventArgs"},
+            {"System.EventArgs`2", "EventArgs"},
+
+            {"Tizen.NUI.EventHandlerWithReturnType", "Tizen.NUI.EventHandlerWithReturnType"},
+            {"Tizen.NUI.EventHandlerWithReturnType`1", "Tizen.NUI.EventHandlerWithReturnType"},
+            {"Tizen.NUI.EventHandlerWithReturnType`2", "Tizen.NUI.EventHandlerWithReturnType"},
+            {"Tizen.NUI.EventHandlerWithReturnType`3", "Tizen.NUI.EventHandlerWithReturnType"},
+
```
Comment (olevett) [inline src/EventBuilder/Cecil/EventTemplateInformation.cs]: Bonus whitespace

File: src/EventBuilder/Cecil/EventTemplateInformation.cs [hunk 3]
```diff
@@ -17,12 +17,34 @@ public static class EventTemplateInformation
...
+
         };
 
-        private static string RenameBogusWinRTTypes(string typeName)
+        private static string RenameBogusTypes(string typeName)
```
Comment (olevett) [inline src/EventBuilder/Cecil/EventTemplateInformation.cs]: 👍

File: src/EventBuilder/Cecil/EventTemplateInformation.cs [hunk 4]
```diff
@@ -43,7 +65,12 @@ private static string GetEventArgsTypeForEvent(EventDefinition ei)
                         type.GenericParameters.Zip(generic.GenericArguments, (name, actual) => new { name, actual })) {
                     var realType = GetRealTypeName(kvp.actual);
 
-                    ret = ret.Replace(kvp.name.FullName, realType);
+                    var temp = ret.Replace(kvp.name.FullName, realType);
```
Comment (olevett) [inline src/EventBuilder/Cecil/EventTemplateInformation.cs]: I'm not sure I quite get this change, shouldn't we be replacing type names for multiple generic type arguments?

File: src/EventBuilder/Platforms/Tizen.cs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+                Log.Debug("Using Tizen.NET {Version} released on {Published}", package.Version, package.Published);
+                Log.Debug("{ReleaseNotes}", package.ReleaseNotes);
+
+                packageManager.InstallPackage(package, ignoreDependencies: true, allowPrereleaseVersions: false);
+            });
+
+            var elmSharp = Directory.GetFiles(packageUnzipPath, "ElmSharp*.dll", SearchOption.AllDirectories);
+
+            foreach (var assembly in elmSharp) {
```
Comment (olevett) [inline src/EventBuilder/Platforms/Tizen.cs]:
```C#
Assemblies.AddRange(elmSharp);
```

File: src/EventBuilder/Platforms/Tizen.cs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+            var retryPolicy = Policy
+                .Handle<Exception>()
+                .WaitAndRetry(
+                    5,
+                    retryAttempt => TimeSpan.FromSeconds(Math.Pow(2, retryAttempt)),
+                    (exception, timeSpan, context) => {
+                        Log.Warning(
+                            "An exception was thrown whilst retrieving or installing {packageName}: {exception}",
+                            _packageName, exception);
```
Comment (olevett) [inline src/EventBuilder/Platforms/Tizen.cs]: The string interpolations here (and ln22 and ln42) need a $ at the start of the string, e.g. ln 58.

**Comment by rookiejava** [comment]
@ghuntley  could you review and merge this?

**Comment by olevett** [comment]
Thanks @rookiejava!

**Pull request merged**